File: mycollab-services-ondemand/src/main/java/com/mycollab/ondemand/module/file/service/impl/S3RawContentServiceImpl.java
package com.mycollab.ondemand.module.file.service.impl;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.*;
import com.mycollab.core.MyCollabException;
import com.mycollab.module.file.service.RawContentService;
import com.mycollab.ondemand.module.file.service.AmazonServiceConfiguration;

import java.io.InputStream;

/**
 * @author MyCollab Ltd.
 * @since 1.0
 */
public class S3RawContentServiceImpl implements RawContentService {

    private AmazonServiceConfiguration amazonServiceConfiguration;

    public S3RawContentServiceImpl() {
        this.amazonServiceConfiguration = AmazonServiceConfiguration.getInstance();
    }

    @Override
    public void saveContent(String objectPath, InputStream stream) {
        AmazonS3 s3client = AmazonS3ClientBuilder.defaultClient();
        try {
            ObjectMetadata metaData = new ObjectMetadata();
            metaData.setCacheControl("max-age=8640000");
            metaData.setContentLength(stream.available());
            PutObjectRequest request = new PutObjectRequest(amazonServiceConfiguration.getBucket(), objectPath, stream, metaData);

            s3client.putObject(request.withCannedAcl(CannedAccessControlList.PublicRead));
        } catch (Exception e) {
            throw new MyCollabException("Can not save s3 object path " + objectPath, e);
        }
    }

    @Override
    public InputStream getContentStream(String objectPath) {
        AmazonS3 s3client = amazonServiceConfiguration.newS3Client();

        try {
            S3Object obj = s3client.getObject(new GetObjectRequest(amazonServiceConfiguration.getBucket(), objectPath));
            return obj.getObjectContent();
        } catch (Exception e) {
            throw new MyCollabException("Can not get s3 resource " + objectPath, e);
        }
    }

    @Override
    public void removePath(String objectPath) {
        AmazonS3 s3client = amazonServiceConfiguration.newS3Client();

        try {
            ObjectListing listObjects = s3client.listObjects(amazonServiceConfiguration.getBucket(), objectPath);
            for (S3ObjectSummary objectSummary : listObjects.getObjectSummaries()) {
                s3client.deleteObject(amazonServiceConfiguration.getBucket(), objectSummary.getKey());
            }
        } catch (Exception e) {
            throw new MyCollabException("Can not remove object path " + objectPath, e);
        }
    }

    @Override
    public void renamePath(String oldPath, String newPath) {
        AmazonS3 s3client = amazonServiceConfiguration.newS3Client();

        try {
            ObjectListing listObjects = s3client.listObjects(amazonServiceConfiguration.getBucket(), oldPath);
            for (S3ObjectSummary objectSummary : listObjects.getObjectSummaries()) {
                String key = objectSummary.getKey();
                String appendPath = key.substring(oldPath.length());
                String newAppPath = newPath + appendPath;

                CopyObjectRequest copyRequest = new CopyObjectRequest(
                        amazonServiceConfiguration.getBucket(), key,
                        amazonServiceConfiguration.getBucket(), newAppPath);
                copyRequest.withCannedAccessControlList(CannedAccessControlList.PublicRead);
                s3client.copyObject(copyRequest);

                DeleteObjectRequest deleteRequest = new DeleteObjectRequest(amazonServiceConfiguration.getBucket(), key);
                s3client.deleteObject(deleteRequest);
            }

        } catch (Exception e) {
            throw new MyCollabException("Can not rename from path " + oldPath + " to " + newPath, e);
        }

    }

    @Override
    public void movePath(String oldPath, String destinationPath) {
        removePath(destinationPath);
        renamePath(oldPath, destinationPath);
    }

    @Override
    public long getSize(String path) {
        AmazonS3 s3client = amazonServiceConfiguration.newS3Client();

        try {
            ObjectListing listObjects = s3client.listObjects(amazonServiceConfiguration.getBucket(), path);
            long size = 0;
            for (S3ObjectSummary objectSummary : listObjects.getObjectSummaries()) {
                size += objectSummary.getSize();
            }

            return size;

        } catch (Exception e) {
            throw new MyCollabException("Can not get size of path " + path, e);
        }
    }
}

File: mycollab-services-ondemand/src/main/java/com/mycollab/ondemand/module/file/service/impl/S3RawContentServiceImpl.kt
package com.mycollab.ondemand.module.file.service.impl

import com.amazonaws.services.s3.AmazonS3ClientBuilder
import com.amazonaws.services.s3.model.*
import com.mycollab.core.MyCollabException
import com.mycollab.module.file.service.RawContentService
import com.mycollab.ondemand.module.file.service.AmazonServiceConfiguration

import java.io.InputStream

/**
 * @author MyCollab Ltd.
 * @since 1.0
 */
class S3RawContentServiceImpl : RawContentService {

    private val amazonServiceConfiguration: AmazonServiceConfiguration = AmazonServiceConfiguration.instance

    override fun saveContent(objectPath: String, stream: InputStream) {
        val s3client = AmazonS3ClientBuilder.defaultClient()
        try {
            val metaData = ObjectMetadata()
            metaData.cacheControl = "max-age=8640000"
            metaData.contentLength = stream.available().toLong()
            val request = PutObjectRequest(amazonServiceConfiguration.bucket, objectPath, stream, metaData)

            s3client.putObject(request.withCannedAcl(CannedAccessControlList.PublicRead))
        } catch (e: Exception) {
            throw MyCollabException("Can not save s3 object path $objectPath", e)
        }

    }

    override fun getContentStream(objectPath: String): InputStream {
        val s3client = amazonServiceConfiguration.newS3Client()

        try {
            val obj = s3client.getObject(GetObjectRequest(amazonServiceConfiguration.bucket, objectPath))
            return obj.objectContent
        } catch (e: Exception) {
            throw MyCollabException("Can not get s3 resource $objectPath", e)
        }

    }

    override fun removePath(objectPath: String) {
        val s3client = amazonServiceConfiguration.newS3Client()

        try {
            val listObjects = s3client.listObjects(amazonServiceConfiguration.bucket, objectPath)
            for (objectSummary in listObjects.objectSummaries) {
                s3client.deleteObject(amazonServiceConfiguration.bucket, objectSummary.key)
            }
        } catch (e: Exception) {
            throw MyCollabException("Can not remove object path $objectPath", e)
        }

    }

    override fun renamePath(oldPath: String, newPath: String) {
        val s3client = amazonServiceConfiguration.newS3Client()

        try {
            val listObjects = s3client.listObjects(amazonServiceConfiguration.bucket, oldPath)
            for (objectSummary in listObjects.objectSummaries) {
                val key = objectSummary.key
                val appendPath = key.substring(oldPath.length)
                val newAppPath = newPath + appendPath

                val copyRequest = CopyObjectRequest(
                        amazonServiceConfiguration.bucket, key,
                        amazonServiceConfiguration.bucket, newAppPath)
                copyRequest.withCannedAccessControlList(CannedAccessControlList.PublicRead)
                s3client.copyObject(copyRequest)

                val deleteRequest = DeleteObjectRequest(amazonServiceConfiguration.bucket, key)
                s3client.deleteObject(deleteRequest)
            }

        } catch (e: Exception) {
            throw MyCollabException("Can not rename from path $oldPath to $newPath", e)
        }

    }

    override fun movePath(oldPath: String, destinationPath: String) {
        removePath(destinationPath)
        renamePath(oldPath, destinationPath)
    }

    override fun getSize(objectPath: String): Long {
        val s3client = amazonServiceConfiguration.newS3Client()

        try {
            val listObjects = s3client.listObjects(amazonServiceConfiguration.bucket, objectPath)
            return listObjects.objectSummaries.map { it.size }.sum()

        } catch (e: Exception) {
            throw MyCollabException("Can not get size of path $objectPath", e)
        }

    }
}
